Manage load balancer creation and deletion asynchronously in a ServiceController

a-robinson opened the pull request:
Closes #5180.

There's still some unit tests to add and a few TODOs to cleanup in the near future, but this is passing all service-related e2es consistently, and should be mergeable as-is. I'll run a full e2e tonight.

cc @lavalamp to make sure I'm using the DeltaFIFO queue correctly.

a-robinson committed:
Put the node controller in its own package rather than in the generic
cloudprovider/controller package.

a-robinson committed:
Implement a ServiceController that watches services and handles keeping
external load balancers up-to-date based on the service's specs, using
the new DeltaFIFO watch queue class. Remove the old registry REST
handler code for creating/updating/deleting load balancers.

Also clean up a bunch of the GCE cloudprovider code related to load balancers.

a-robinson committed:
Fix the services namespace test to wait for the services' load balancers
to be created.

a-robinson committed:
Remove the cloud provider field from the services REST handler and the master
now that load balancers are handled by the ServiceController.

lavalamp commented on pkg/cloudprovider/servicecontroller/servicecontroller.go:
```diff
@@ -0,0 +1,407 @@
+	lw := cache.NewListWatchFromClient(s.kubeClient.(*client.Client), "services", api.NamespaceAll, fields.Everything())
+	// TODO: Add a compressor that intelligently squashes together updates.
+	serviceQueue := cache.NewDeltaFIFO(cache.MetaNamespaceKeyFunc, nil, s)
+	cache.NewReflector(lw, &api.Service{}, serviceQueue, 0).Run()
```
Do services need occasional reprocessing? If so, suggest a non-zero poll period.

a-robinson replied:
I don't believe so, other than maybe as a safety mechanism? Is that the best practice for our various controllers?

lavalamp commented on pkg/cloudprovider/servicecontroller/servicecontroller.go:
```diff
@@ -0,0 +1,407 @@
+	for {
+		deltas, ok := serviceQueue.Pop().(cache.Deltas)
+		if !ok {
+			glog.Errorf("Received object from service watcher that wasn't Deltas: %+v", deltas)
```
Suggest doing the pop and type assertion in two steps if you want a useful debug message. But probably doesn't matter since DeltaFIFO isn't capable of giving you anything else.

a-robinson replied:
Good point, changed.

lavalamp commented on pkg/cloudprovider/servicecontroller/servicecontroller.go:
```diff
@@ -0,0 +1,407 @@
+
+// ListKeys implements the interface required by DeltaFIFO to list the keys we
+// already know about.
+func (s *ServiceController) ListKeys() []string {
```
Consider if you want this to be public-- you may want to make a (private) accessor type for providing this to DeltaFIFO.

a-robinson replied:
I hadn't noticed the KeyListerFunc type initially, that should do the trick for wrapping this in a closure.

a-robinson commented on pkg/cloudprovider/servicecontroller/servicecontroller.go:
```diff
@@ -0,0 +1,407 @@
+		if !ok {
+			glog.Errorf("Received object from service watcher that wasn't Deltas: %+v", deltas)
+		}
+		if len(deltas) != 1 {
```
I realized that this is wrong -- I'm fixing this so that we do process all deltas until I can add a compressor.

lavalamp replied:
You aren't actually passing a compressor function, so I don't expect this to work out very well for you... :)

a-robinson replied:
Yup, I noticed that too :) I initially fixed by processing all of the deltas, but I think we only need the most recent state so long as delete is idempotent.

lavalamp commented on pkg/cloudprovider/servicecontroller/servicecontroller.go:
```diff
@@ -0,0 +1,407 @@
+			glog.Errorf("Ignoring service Deltas that weren't properly compressed: %+v", deltas)
+			continue
+		}
+		service, ok := deltas[0].Object.(*api.Service)
```
If you only care about the most recent update you can use deltas.Newest().Object...

a-robinson replied:
That should work, thanks.

lavalamp commented:
Awesome! I only reviewed the use of DeltaFIFO-- super happy to see it getting used already.

zmerlynn commented on cluster/saltbase/salt/kube-controller-manager/default:
```diff
@@ -17,6 +17,9 @@
 {% if pillar['node_instance_prefix'] is defined -%}
   {% set minion_regexp = "--minion_regexp='" + pillar['node_instance_prefix'] + ".*'" -%}
 {% endif -%}
```
You may want be pedantic and
```
{% set cluster_name = "" -%}
```

here.

(I just tuned in for the Salt change to see if it was going to conflict badly in the same file.)

a-robinson replied:
Thanks, done.

a-robinson commented:
Thanks for the feedback, @lavalamp!

@quinton-hoole, feel free to reassign if you don't have time for this or think someone else would be a better choice. I know you've got a lot on your plate. :)

quinton-hoole commented:
My apologies for the delay.  I'll get onto this right away.

a-robinson commented:
Updated and rebased on top of master.

lavalamp commented:
DeltaFIFO usage LGTM. You might want to squish a few commits when you're done. :)

quinton-hoole commented on cmd/kubernetes/kubernetes.go:
```diff
@@ -126,11 +127,14 @@ func runControllerManager(machineList []string, cl *client.Client, nodeMilliCPU,
+	kubeletClient := &client.HTTPKubeletClient{Client: http.DefaultClient, Port: ports.KubeletPort}
 
-	nodeController := nodeControllerPkg.NewNodeController(nil, "", machineList, nodeResources, cl, kubeClient, 10, 5*time.Minute)
+	nodeController := nodecontroller.NewNodeController(nil, "", machineList, nodeResources, cl, kubeletClient, 10, 5*time.Minute)
```
Nit: Changing these literals to named constants would make the code clearer.

a-robinson replied:
Cleaned up.

quinton-hoole commented on pkg/cloudprovider/gce/gce.go:
```diff
@@ -211,10 +217,21 @@ func (gce *GCECloud) waitForRegionOp(op *compute.Operation, region string) error
 func (gce *GCECloud) TCPLoadBalancerExists(name, region string) (bool, error) {
 	_, err := gce.service.ForwardingRules.Get(gce.projectID, region, name).Do()
+	if err == nil {
+		return true, err
```
nit: return true, nil

a-robinson replied:
Done.

a-robinson commented:
Thanks for the suggestions! Fixed all of them, squashed commits together to where I think they should end up, and rebased on top of master.

a-robinson commented:
Also cc @davidopp since the first commit of this PR touches the nodecontroller, although it's only a directory move.

quinton-hoole commented on pkg/cloudprovider/gce/gce.go:
```diff
@@ -251,18 +273,17 @@ func (gce *GCECloud) CreateTCPLoadBalancer(name, region string, externalIP net.I
 		Name:       name,
 		IPProtocol: "TCP",
 		PortRange:  fmt.Sprintf("%d-%d", minPort, maxPort),
-		Target:     pool,
```
It's existing code, but I couldn't help noticing that the above for loop will end up creating a single forwarding rule spanning everything from minPort to maxPort (I think).  e.g. if ports[0]==1 and ports[1]==65000 we'll end up forwarding 65000 ports, and not just 2.  Am I right?

quinton-hoole replied:
If I am right, please file an issue to fix the bug.

a-robinson replied:
Yup, that is right. This was recently added with the multi-port work, so I'm not sure how much it was thought through, but PortRange is the only way to specify ports in a GCE forwarding rule.

Since each forwarding rule has its own IP and there's no way to specify multiple smaller ranges, I don't believe this is a bug. The iptables rules the proxy creates will only accept packets targeted to that IP that are actually one on the service's ports - others won't be handled by the iptables rules.

a-robinson replied:
@thockin in case he has more context on this, since he wrote it.

bgrant0607 replied:
Correct, not ideal, but not a bug.

thockin replied:
yep

quinton-hoole commented on pkg/cloudprovider/servicecontroller/servicecontroller.go:
```diff
@@ -0,0 +1,426 @@
+func (s *ServiceController) Run() {
+	if err := s.init(); err != nil {
+		glog.Errorf("Failed to initialized ServiceController: %v", err)
+		return
```
This swallows the error, making it indistinguishable from success to the caller.  Does anything other than a log message happen when this error occurs?

quinton-hoole replied:
I guess that ultimately this needs to bubble up to the API, ideally as a failure to create the service, or as a service with a visibly broken external load balancer.

quinton-hoole commented on pkg/cloudprovider/servicecontroller/servicecontroller.go:
```diff
@@ -0,0 +1,426 @@
+	// for some reason.
+	if _, ok := s.kubeClient.(*client.Client); !ok {
+		glog.Error("ServiceController only works with real Client objects, but was passed something else satisfying the client Interface.")
+		return
```
Same comment as above.

quinton-hoole commented on pkg/cloudprovider/servicecontroller/servicecontroller.go:
```diff
@@ -0,0 +1,426 @@
+		if err != nil {
+			// Add the failed service back to the queue so we'll retry it.
+			glog.Error(err)
+			time.Sleep(5 * time.Second)
```
Having an arbitrarily long sleep like this buried deep in your code like this can be dangerous.   Let me go through the rest of your code before proposing alternatives.

smarterclayton replied:
Retry and waits should be handled outside this loop - I believe lavalamp was in the process of establishing a generic controller framework for retryable errors based in part on work @ironcladlou had done for Openshift.  Don't know how far that has progressed.  If it's lower in the queue we can push on that a bit more (Ben recently added qps limits on retries to our version which prevents hot failure loops)

quinton-hoole commented on pkg/cloudprovider/servicecontroller/servicecontroller.go:
```diff
@@ -0,0 +1,426 @@
+	}
+	if cached {
+		// If the service already exists but needs to be updated, delete it so that
+		// we can recreate it cleanly.
```
Won't this create an unnecessarily long break in network connectivity to the service?  Can it not rather be updated in place?

a-robinson replied:
That's a very fair point. Unfortunately, GCE forwarding rules can't be updated in place other than to point them at a different target pool.

However, we can break down the problem a bit. The reasons we may need to update an external load balancer (as you can determine from the needsUpdate function) are if (a) the public IPs changed, or (b) the ports changed.

There are a couple alternative ways to handle these that might be worth considering. For changing ports, we could:
1. Create a new forwarding rule pointed at the same target pool with the new ports, then update the REST object with the new IP address this would create, then delete the old one. The downside is that this would require changing the IP and tracking the names of the forwarding rules rather than knowing exactly what the name could be.
2. Always set the port range to be 1-65535, regardless of which ports the service actually uses. Then we can disregard changes to the port numbers.
3. Accept a bit of downtime, but at least keep the IP around. Reserve the IP the FW is using, then delete the old FW, then create the new one with the correct IP, then un-reserve the IP. However, reserving IPs is GCE-specific, so it'd have to go in the GCE cloud provider implementation of create and update load balancer.

For changing IP addresses, we could:
1. The same as option 1 above.
2. Put the IP address in the name of the load balancer, so that we could easily create the new one while the old one still exists while still having deterministic names. It'd be tricky to do this for new load balancers, though.

For now, accepting a bit of downtime on service update doesn't seem absurd, but we could try one of these other options as a future issue.

Really, it seems like IP addresses chosen by the cloud provider should be in the service's status, not the service's spec, since this code can't determine whether the IP address was user-specified and needs to be respected vs whether it was arbitrarily chosen by the cloud provider when the previous balancer was created. Do you or @thockin think this sort of change is feasible given the recent switch to v1beta3?

quinton-hoole commented on pkg/cloudprovider/servicecontroller/servicecontroller.go:
```diff
@@ -0,0 +1,426 @@
+		}
+		glog.Warningf("Failed to persist updated PublicIPs to service %s after creating its external load balancer: %v",
+			service.Name, err)
+		time.Sleep(clientRetryInterval)
```
How do you avoid the situation where the rate of change of services exceeds the rate at which you are able to update the external load balancers?  Given the serial way in which you do updates (if I'm reading the code correctly), and the number of places in which you sleep for seconds at a time, I'm concerned.

a-robinson replied:
Yeah, this is pretty terrible. I'll add in the parallelism now, such that each service can be processed separately from each other, but changes to an individual service will be done serially.

quinton-hoole commented on pkg/cloudprovider/servicecontroller/servicecontroller.go:
```diff
@@ -0,0 +1,426 @@
+		// This is panic-worthy, since the queue shouldn't have been able to
+		// handle the service if it couldn't generate a name for it.
+		glog.Errorf("Couldn't generate namespaced name for service: %v", err)
+		return nil
```
Yes, you should not swallow the error like this, making it invisible to the caller.

a-robinson replied:
I had intended for the error to indicate whether the failure should be retried, but I suppose that may be better handled by returning two values. I'll change to that model, with one bool indicating whether to retry and one error indicating if it succeeded.

quinton-hoole commented on pkg/cloudprovider/servicecontroller/servicecontroller.go:
```diff
@@ -0,0 +1,426 @@
+	return nil
+}
+
+func (s *ServiceController) deleteExternalLoadBalancer(service *api.Service) error {
```
nit: It's good that you're doing idempotent deletes, but from a naming point of view this is not technically a delete function.   Consider renaming it to something more like (ensureDeleted...).

a-robinson replied:
Done.

quinton-hoole commented on pkg/cloudprovider/servicecontroller/servicecontroller.go:
```diff
@@ -0,0 +1,426 @@
+	keys := make([]string, len(s.serviceMap))
+	i := 0
+	for k := range s.serviceMap {
+		keys[i] = k
```
nit: I think that you can use append() here, and get rid of i entirely.

a-robinson replied:
If I also specify that the initial length of the slice is 0, yes. Done.

quinton-hoole commented on pkg/cloudprovider/servicecontroller/servicecontroller.go:
```diff
@@ -0,0 +1,426 @@
+}
+
+func (s *ServiceController) loadBalancerName(service *api.Service) string {
+	return s.clusterName + "-" + service.Namespace + "-" + service.Name
```
I noticed recently that this can quite exceed the name length limits on GCE, and probably other cloud providers too.  Consider using a shorter hash instead, that's crafted to fit within the limits?

a-robinson replied:
Added a TODO. I don't want this to get lost in the larger PR, especially since it's something that's would affect PR #5225, which is still in-flight.

quinton-hoole commented on pkg/cloudprovider/servicecontroller/servicecontroller.go:
```diff
@@ -0,0 +1,426 @@
+	if len(xPorts) != len(yPorts) {
+		return false
+	}
+	for i := range xPorts {
```
Do you need to consider identical sets that are ordered differently?  Perhaps sort before comparing?

a-robinson replied:
I just pulled this code from the REST handler, but that's a fair point, since it doesn't look like our validation functions require them to be sorted.

Since go's sorting functions are in-place and I'd rather not modify the user-specified order, I just used a map for checking the equality.

quinton-hoole commented:
OK, I've done a first pass review.  High level comment is that the serial loop through all services, which includes synchronous callouts to potentially flakey cloud providers, with multiple sleeps and retries seems dangerous.  We should try to figure out a more robust mechanism, perhaps with a configurable sized pool of go routines feeding off a channel.

smarterclayton commented on pkg/registry/service/rest.go:
```diff
@@ -97,6 +94,14 @@ func (rs *REST) Create(ctx api.Context, obj runtime.Object) (runtime.Object, err
+	// even though it'll be created by the ServiceController.
+	if service.Spec.CreateExternalLoadBalancer {
+		if _, err := getTCPPorts(service); err != nil {
+			return nil, err
```
This should be handled in validation.

a-robinson replied:
Doh, yup, I'll move it there. Thanks!

fgrzadkowski commented:
cc-ing @jszczepkowski who wanted to add logic to update GCE targetPools when nodes changes. I think that service controller is the right place to do it.

a-robinson committed:
Move validation of load balancers only supporting TCP ports to validation.go.

quinton-hoole commented:
a-robinson@ please let me know when you want me to re-review.  Until then, I'm re-assigning this PR to you.

a-robinson committed:
Do service creation/update/deletion work in a pool of goroutines, protecting
each service with a lock to ensure that no two goroutines will process a
service at the same time. This is needed to avoid weird race conditions.

a-robinson commented:
Sorry for the delay @quinton-hoole! I've been out the last few days.

I've rebased on top of head and updated this to respond to all comments. In particular, I made the service reconciliation tasks run in a pool of goroutines while being very careful to avoid the races that could come up from processing the same service in two goroutines simultaneously.

a-robinson commented:
Ping?

quinton-hoole commented:
Yes, sorry - I've been otherwise occupied.  I'l try to finish this review today.

quinton-hoole commented on pkg/cloudprovider/servicecontroller/servicecontroller.go:
```diff
@@ -0,0 +1,468 @@
+}
+
+// TODO: Use a shorter name that's less likely to be longer than cloud
+// providers' length limits.
```
Please file an issue for this, and add a link to it here.

a-robinson replied:
Created #6812 and added it to the TODO.

quinton-hoole commented on pkg/cloudprovider/servicecontroller/servicecontroller.go:
```diff
@@ -0,0 +1,468 @@
+		return false
+	}
+	// Use a map for comparison since port slices aren't necessarily sorted.
+	xPortMap := make(map[int]bool)
```
nit: It's probably neater to just sequentially compare a a sorted copy of xPorts against a sorted copy of yPorts.  Up to you.

a-robinson replied:
The length check is already there, a couple lines above.

Comparing sorted lists would be neater, but it requires sorting them first, since I don't know of any restriction that a service's ports have to be listed in increasing order.

quinton-hoole commented on pkg/cloudprovider/servicecontroller/servicecontroller.go:
```diff
@@ -0,0 +1,468 @@
+			time.Sleep(5 * time.Second)
+			serviceQueue.AddIfNotPresent(deltas)
+		} else if err != nil {
+			glog.Errorf("Failed to process service delta. Not retrying: %v", err)
```
You're still just silently logging errors here. Short of digging through logs and noticing the error, the user won't know that an error has occurred, or that you've given up trying to build a working service for her. I think that you need to have the service go into some sort of error state in this case.

smarterclayton replied:
Always use util.HandleError() if you are eating an error (in general don't eat errors, but when you have to that's what HandleError is for).

> On Apr 13, 2015, at 6:37 PM, Quinton Hoole <notifications@github.com> wrote:
> 
> In pkg/cloudprovider/servicecontroller/servicecontroller.go:
> 
> > +		if !ok {
> > +			glog.Errorf("Received object from service watcher that wasn't Deltas: %+v", newItem)
> > +		}
> > +		delta := deltas.Newest()
> > +		if delta == nil {
> > +			glog.Errorf("Received nil delta from watcher queue.")
> > +			continue
> > +		}
> > +		err, shouldRetry := s.processDelta(delta)
> > +		if shouldRetry {
> > +			// Add the failed service back to the queue so we'll retry it.
> > +			glog.Errorf("Failed to process service delta. Retrying: %v", err)
> > +			time.Sleep(5 * time.Second)
> > +			serviceQueue.AddIfNotPresent(deltas)
> > +		} else if err != nil {
> > +			glog.Errorf("Failed to process service delta. Not retrying: %v", err)
> You're still just silently logging errors here. Short of digging through logs and noticing the error, the user won't know that an error has occurred, or that you've given up trying to build a working service for her. I think that you need to have the service go into some sort of error state in this case.
> 
> —
> Reply to this email directly or view it on GitHub.
>

a-robinson replied:
Thanks, Clayton! Switched to using util.HandleError().

The only places that this code returns a notRetryable error are from programming bugs (deltaFIFO returning something invalid) or from being unable to persist publicIPs to the apiserver. The latter is important, but I'm hesitant to always retry on that because I'm worried that I missed handling some class of non-retryable errors in persistUpdate, which could cause us to continue retrying it forever. This could admittedly cause an issue during an apiserver outage, but when reconnecting after an apiserver outage we should receive a sync of all the services anyways.

quinton-hoole commented on pkg/cloudprovider/servicecontroller/servicecontroller.go:
```diff
@@ -0,0 +1,468 @@
+// indicator of whether the processing should be retried.
+func (s *ServiceController) processDelta(delta *cache.Delta) (error, bool) {
+	service, ok := delta.Object.(*api.Service)
+	namespacedName := ""
```
I'd much prefer the use of types.NamespacedName here, rather than a bland string as the cache key.  We've seen many bugs involving missing namespaces, that that will catch.

a-robinson replied:
Ah, I didn't realize that existed, thanks!

quinton-hoole commented on pkg/cloudprovider/servicecontroller/servicecontroller.go:
```diff
@@ -0,0 +1,468 @@
+			return nil
+		}
+		if errors.IsNotFound(err) {
+			glog.Infof("Not persisting update to service that no longer exists: %v", err)
```
You're masking the error here again buy just logging it an returning success. You need to return an error so that the caller can reget and retry to update against the new version.

a-robinson replied:
What's the error? Unless If the service no longer exists, it must have been deleted. If it was deleted, we don't want to re-create it, we just want to bail out so that we can process the deletion event. I've tried explaining this in a new comment, but please tell me if I'm missing something or this isn't clear.

quinton-hoole replied:
Good point.  Agreed.

quinton-hoole commented:
OK, I think once we have this round of comments resolved we're in good shape to merge.

quinton-hoole commented:
PS:  Apologies for the duplicate comments - I kept forgetting and then reremembering that comments in the commits as opposed to the files get thrown away by Git/GitHub.

a-robinson commented:
Rebased on top of master (mostly had issues with #6292) and squashed down to a core set of commits.

quinton-hoole commented:
Shippable and Travis say that your code doesn't compile?

"# github.com/GoogleCloudPlatform/kubernetes/pkg/cloudprovider/servicecontroller
_output/local/go/src/github.com/GoogleCloudPlatform/kubernetes/pkg/cloudprovider/servicecontroller/servicecontroller_test.go:92: cannot use "foo/bar" (type string) as type types.NamespacedName in argument to controller.createLoadBalancerIfNeeded
FAIL	github.com/GoogleCloudPlatform/kubernetes/pkg/cloudprovider/servicecontroller [build failed]"

a-robinson commented:
Heh, I ran the unit tests before rebasing and the end-to-end tests afterward, but not the unit tests afterward. Sorry! Should be good now.

quinton-hoole commented:
LGTM.

thockin commented:
LGTM - thanks for taking this one on - it was pretty daunting.
